refactor(devtools): Make every ng function optional in the typings

JeanMeche opened the pull request:
This will help in the future guarding against old apps what don't have the new APIs implemented

AleksanderBodurri commented on devtools/projects/ng-devtools-backend/src/lib/ng-debug-api/ng-debug-api.ts:
```diff
@@ -16,12 +16,14 @@ import type {ɵGlobalDevModeUtils as GlobalDevModeUtils} from '@angular/core';
+ * Type guard that checks whether a given debug API is supported within window.ng
  *
- * @returns boolean
+ * @returns wether the ng object includes the given debug API
```
nit: Typo `whether"

AleksanderBodurri commented on packages/core/src/render3/util/global_utils.ts:
```diff
@@ -82,6 +82,21 @@ const globalUtilsFunctions = {
 };
+
+// Those functions are considered available in any app context and never require a fallback.
+type AlwaysAvailableFns =
```
Some of these functions are newer than getDirectives which is listed as optional here.  I think only `getComponent` and `getOwningComponent` should be considered non-optional here. In the future we may consider making everything "optional" and feature sniffing all API usages on the DevTools side

JeanMeche replied:
Then maybe let's make it all optional like Doug did but remove any type assertions ?

dgp1130 replied:
I'd lean towards just leaving everything optional for now.

I agree something like `getComponent` is _probably_ safe to assume is present all the time, but expanding support to Wiz might involve resetting some of these kinds of assumptions.

AleksanderBodurri replied:
Context for future readers: there is no plans for these APIs to become unsupported in future versions of Angular, only that Angular DevTools may begin to treat them as optional because they may not be present in Wiz applications

JeanMeche committed:
refactor(devtools): Make every ng function optional in the typings

This will help in the future guarding against old apps what don't have the new APIs implemented

dgp1130 reviewed (approved):
Thanks for taking this a step further!

AndrewKushnir commented:
This PR was merged into the repository by commit a94c912a3a5126a07b4d628b7dd3b634fa2471a0.

The changes were merged into the following branches: main
